Can't compile with Rust 1.39

With rust 1.39 is not possible to compile this crate anymore, I get the followig errors. Do you have any ideas?

```
error[E0369]: binary operation `==` cannot be applied to type `__mbstate_t__bindgen_ty_1`
    --> /home/andrea/Workspace/git/video_ludo/target/debug/build/ffmpeg-sys-77d9e656be992a77/out/bindings.rs:3000:5
     |
3000 |     pub __value: __mbstate_t__bindgen_ty_1,
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     |
     = note: an implementation of `std::cmp::PartialEq` might be missing for `__mbstate_t__bindgen_ty_1`

error[E0369]: binary operation `!=` cannot be applied to type `__mbstate_t__bindgen_ty_1`
    --> /home/andrea/Workspace/git/video_ludo/target/debug/build/ffmpeg-sys-77d9e656be992a77/out/bindings.rs:3000:5
     |
3000 |     pub __value: __mbstate_t__bindgen_ty_1,
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     |
     = note: an implementation of `std::cmp::PartialEq` might be missing for `__mbstate_t__bindgen_ty_1`

error[E0277]: the trait bound `__mbstate_t__bindgen_ty_1: std::cmp::Eq` is not satisfied
    --> /home/andrea/Workspace/git/video_ludo/target/debug/build/ffmpeg-sys-77d9e656be992a77/out/bindings.rs:3000:5
     |
3000 |     pub __value: __mbstate_t__bindgen_ty_1,
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::cmp::Eq` is not implemented for `__mbstate_t__bindgen_ty_1`
     |
     = note: required by `std::cmp::AssertParamIsEq`
```

After a `git bisect`, it looks like this issue first appears in 8a6d9c6eecddc4c95c63977378c490196c19ea45, in which bindgen was updated from 0.30 to 0.31. These were all run with the current stable Rust compiler (1.39.0).

It appears that versions of bindgen between 0.31 and 0.51.0 interact with a compiler behavior change in 1.39 -- see https://github.com/rust-lang/rust-bindgen/pull/1627 and the related discussion at https://github.com/rust-lang/rust/issues/64710. The solution is to upgrade bindgen to the latest fixed version at 0.51.1. This is accomplished in PR #79.

However, even with the upgrade, #79 does not fix all issues (at least on my machine 😄), producing outputs like this:

```
error[E0428]: the name `FP_ZERO` is defined multiple times
    --> /home/daniel/rust-ffmpeg-sys/target/debug/build/ffmpeg-sys-790f9f63289709f5/out/bindings.rs:2855:1
     |
572  | pub const FP_ZERO: libc::c_int = 2;
     | ----------------------------------- previous definition of the value `FP_ZERO` here
...
2855 | pub const FP_ZERO: _bindgen_ty_1 = _bindgen_ty_1::FP_ZERO;
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `FP_ZERO` redefined here
     |
     = note: `FP_ZERO` must be defined only once in the value namespace of this module
```

[build.rs line 906](https://github.com/meh/rust-ffmpeg-sys/blob/4f14151b9b8134f1f029d49d02cbea5c7337dedb/build.rs#L906-L910) attempts to fix this by blacklisting the types involved (`FP_NAN`, `FP_INFINITE`, `FP_ZERO`, `FP_SUBNORMAL`, and `FP_NORMAL`). However, at least on my machine, this is insufficient. The [recommended way](https://github.com/rust-lang/rust-bindgen/issues/687#issuecomment-388277405) of handling this is by implementing `will_parse_macro` and returning `Ignore` for any of the macros given here. This replaces the type blacklisting, allowing that code to be removed.

I will work up a PR and check that it builds on Travis as well as my machine.

--- a/.travis.yml
+++ b/.travis.yml
@@ -15,6 +15,9 @@ addons:
     packages:
       - build-essential
 before_install:
+  # Without rustfmt, bindgen puts everything on one line and any warnings dump so many logs they break Travis
+  # See https://github.com/rust-lang/rust-bindgen/issues/1600
+  - rustup component add rustfmt
   - if [[ $TRAVIS_OS_NAME == 'linux' ]]; then ./.travis/install_linux.sh; fi
   - if [[ $TRAVIS_OS_NAME == 'osx' ]]; then brew update; fi
   - if [[ $TRAVIS_OS_NAME == 'osx' ]]; then brew install yasm; fi

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -18,7 +18,7 @@ libc = "0.2"
 num_cpus = "1.0"
 cc = "1.0"
 pkg-config = "0.3"
-bindgen    = "0.32"
+bindgen    = "0.51.1"
 regex      = "0.2"
 
 [features]

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -11,6 +11,10 @@ description = "FFI bindings to FFmpeg"
 repository  = "https://github.com/meh/rust-ffmpeg-sys"
 keywords    = ["audio", "video"]
 
+[lib]
+# Disable doctests as a workaround for https://github.com/rust-lang/rust-bindgen/issues/1313
+doctest = false
+
 [dependencies]
 libc = "0.2"
 

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -15,11 +15,11 @@ keywords    = ["audio", "video"]
 libc = "0.2"
 
 [build-dependencies]
-num_cpus = "1.0"
-cc = "1.0"
+num_cpus   = "1.11"
+cc         = "1.0"
 pkg-config = "0.3"
 bindgen    = "0.51.1"
-regex      = "0.2"
+regex      = "1.3"
 
 [features]
 default  = ["avcodec", "avdevice", "avfilter", "avformat", "swresample", "swscale"]

--- a/build.rs
+++ b/build.rs
@@ -12,7 +12,7 @@ use std::process::Command;
 use std::str;
 
 use regex::Regex;
-use bindgen::callbacks::{IntKind, ParseCallbacks};
+use bindgen::callbacks::{IntKind, ParseCallbacks, MacroParsingBehavior};
 
 #[derive(Debug)]
 struct Library {
@@ -69,6 +69,20 @@ impl ParseCallbacks for IntCallbacks {
             Some(IntKind::Int)
         } else {
             None
+        }
+    }
+
+    // https://github.com/rust-lang/rust-bindgen/issues/687#issuecomment-388277405
+    fn will_parse_macro(&self, name: &str) -> MacroParsingBehavior {
+        use MacroParsingBehavior::*;
+
+        match name {
+            "FP_INFINITE" => Ignore,
+            "FP_NAN" => Ignore,
+            "FP_NORMAL" => Ignore,
+            "FP_SUBNORMAL" => Ignore,
+            "FP_ZERO" => Ignore,
+            _ => Default,
         }
     }
 }
@@ -902,12 +916,6 @@ fn main() {
     let mut builder = bindgen::Builder::default()
         .clang_args(clang_includes)
         .ctypes_prefix("libc")
-        // https://github.com/servo/rust-bindgen/issues/687
-        .blacklist_type("FP_NAN")
-        .blacklist_type("FP_INFINITE")
-        .blacklist_type("FP_ZERO")
-        .blacklist_type("FP_SUBNORMAL")
-        .blacklist_type("FP_NORMAL")
         // https://github.com/servo/rust-bindgen/issues/550
         .blacklist_type("max_align_t")
         .rustified_enum("*")

--- a/build.rs
+++ b/build.rs
@@ -117,16 +117,14 @@ fn search() -> PathBuf {
 }
 
 fn fetch() -> io::Result<()> {
-    let status = try!(
-        Command::new("git")
+    let status = Command::new("git")
             .current_dir(&output())
             .arg("clone")
             .arg("-b")
             .arg(format!("release/{}", version()))
             .arg("https://github.com/FFmpeg/FFmpeg")
             .arg(format!("ffmpeg-{}", version()))
-            .status()
-    );
+            .status()?;
 
     if status.success() {
         Ok(())
@@ -281,24 +279,22 @@ fn build() -> io::Result<()> {
     }
 
     // run make
-    if !try!(
-        Command::new("make")
+    if !Command::new("make")
             .arg("-j")
             .arg(num_cpus::get().to_string())
             .current_dir(&source())
-            .status()
-    ).success()
+            .status()?
+            .success()
     {
         return Err(io::Error::new(io::ErrorKind::Other, "make failed"));
     }
 
     // run make install
-    if !try!(
-        Command::new("make")
+    if !Command::new("make")
             .current_dir(&source())
             .arg("install")
-            .status()
-    ).success()
+            .status()?
+            .success()
     {
         return Err(io::Error::new(io::ErrorKind::Other, "make install failed"));
     }

